src/EditorFeatures/VisualBasic/LineCommit/CommitBufferManager.vb

```diff
@@ -97,63 +96,119 @@ Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.LineCommit
+                                  End Sub
+                    AddHandler textBuffer.Changing, handler
+
+                    Try
+                        ' Start to suppress commits to ensure we don't have any sort of re-entrancy in this process.
+                        ' We've seen bugs (17015) where waits triggered by some computation might re-enter.
+                        Using BeginSuppressingCommits()
+                            ' It's possible that an edit may already be in progress. In this scenario, there's
+                            ' really nothing we can do, so we'll just skip the format
+                            If _buffer.EditInProgress Then
+                                Return
+                            End If
+
+                            Dim dirtyRegion = _dirtyState.DirtyRegion.GetSpan(_buffer.CurrentSnapshot)
+                            Dim info As FormattingInfo
+                            If Not TryComputeExpandedSpanToFormat(dirtyRegion, info, cancellationToken) Then
+                                Return
+                            End If
+
+                            Dim useSemantics = info.UseSemantics
+                            If useSemantics AndAlso Not isExplicitFormat Then
+                                ' Avoid using semantics for formatting extremely large dirty spans without an explicit request
+                                ' from the user. The "large span threshold" is 7000 lines. The 7000 line threshold is an
+                                ' estimated value accounting for a lower-bound of the algorithmic complexity of text
+                                ' differencing in designer cases along with measurements of a pathological example demonstrated
+                                ' at 14000 lines. We expect Windows Forms designer formatting operations to run in under ~15
+                                ' seconds on average current hardware when nearing the threshold.
+                                Dim startLineNumber = 0
+                                Dim startCharIndex = 0
+                                Dim endLineNumber = 0
+                                Dim endCharIndex = 0
+                                info.SpanToFormat.GetLinesAndCharacters(startLineNumber, startCharIndex, endLineNumber, endCharIndex)
+                                If endLineNumber - startLineNumber > 7000 Then
+                                    useSemantics = False
+                                End If
+                            End If
+
+                            ' Get the original tree, but come back to the UI thread so we can capture the current buffer snapshot
+                            Dim tree = If(blocking,
+                                _dirtyState.BaseDocument.GetSyntaxTreeSynchronously(cancellationToken),
+                                Await _dirtyState.BaseDocument.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(True))
```

Comment:
this will show up in a few places.  we either continue calling GetFooSynchronously, if this is the blocking case, or we do full async otherwise.